ch4/ofi: Improved noncontiguous put/get

raffenet opened the pull request:
## Pull Request Description

Noncontiguous RMA has several issues regarding datatype optimization and correctness. This PR removes the old, complex native implementation entirely. A new native implementation is added for when origin and target datatypes are both high density. See #4468.

<!-- AUTHOR: After creating this merge request, check off each of the following items as you complete them. -->

## Expected Impact

Simpler code. Improved performance for many RMA use-cases.

## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [x] Passes warning tests
* [ ] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [x] You or your company has a signed contributor's agreement on file with Argonne
* [x] For non-Argonne authors, request an explicit comment from your companies PR approval manager

raffenet commented:
test:jenkins/ch4/ofi

raffenet commented:
test:mpich/ch4/ofi

raffenet commented:
test:jenkins/ch4/ofi

pavanbalaji commented:
@raffenet While reviewing the `ch4` code, I noticed that the AM layer doesn't perform packing either.  It constructs IOVs in all cases too.  It's a bit better than the `ch4:ofi` code because it constructs the full IOV instead of one element at a time, but that's not optimal either.

Should this PR be expanded to improve the CH4 AM code too?  Or perhaps that should be a separate PR?

raffenet commented:
> @raffenet While reviewing the `ch4` code, I noticed that the AM layer doesn't perform packing either. It constructs IOVs in all cases too. It's a bit better than the `ch4:ofi` code because it constructs the full IOV instead of one element at a time, but that's not optimal either.
> 
> Should this PR be expanded to improve the CH4 AM code too? Or perhaps that should be a separate PR?

I would like to keep any AM code changes separate from this PR. I believe @hzhou might already have a PR to address some of the AM code issues.

hzhou commented:
> I would like to keep any AM code changes separate from this PR. I believe @hzhou might already have a PR to address some of the AM code issues.

I agree that we should keep the PR separate. A likely arrangement is to hold this PR until the AM PR worked out and gets merged first.

That assumes we have a solution for ch4 AM. Let's discuss it --

For ch4 AM RMA, what we needed is to design a new protocol. The current protocol is to send an "IOV" packet in cases of non-contig `put/get`. What is the alternative solution? More specifically, how do we let the remote end know the datatype of another process?

hzhou commented:
Maybe of potential interest, I just created this draft PR #4487, essentially enable `DTPools` to recreate the datatype from a dtpools-description-string. It may be feasible in mpich to have this "datatype-string" conversion routines then we can communicate with remote end using datatype description string in stead of `IOV` arrays.

raffenet commented:
test:mpich/ch4/ofi

raffenet commented:
test:mpich/ch4/ofi

raffenet commented:
test:mpich/ch4/ofi

pavanbalaji commented:
@raffenet and @hzhou   If you are following the discussion with Intel, this PR has become high priority now because of yaksa's dependency on it.  So whatever we decide for this PR and the AM PR would need to be soon.

pavanbalaji commented:
> EDIT2: if yaksa provide a serialization api, then we can use that too.

Yaksa and dataloop (and typerep) provide serialization APIs.  This a core part of the datatype management code.  There's no reason to pass around strings.

raffenet commented:
test:jenkins/ch4/ofi

raffenet commented:
test:jenkins/ch4/ofi

raffenet commented:
test:mpich/ch4/ofi

raffenet commented:
test:jenkins/ch4/ofi

raffenet commented:
test:mpich/ch4/ofi

raffenet commented:
test:mpich/ch4/most

hzhou commented on src/mpid/ch4/src/ch4_impl.h:
```diff
@@ -320,6 +320,19 @@ MPL_STATIC_INLINE_PREFIX void MPIDIG_win_hash_clear(MPIR_Win * win)
+            MPIR_Datatype_get_ptr((datatype_), dt_ptr_);        \
+            MPI_Aint size_;                                     \
+            MPIR_Datatype_get_size_macro((datatype_), size_); \
+            (density_) = size_ / (dt_ptr_)->max_contig_blocks;	\
```
What is `max_contig_blocks`? Is it the same as `MPIR_Typerep_get_iov_len`?

hzhou replied:
I think we should put this macro in `mpir_datatype.h` now, because the context/concept for this code is independent of ofi device or any device. It's a datatype macro and I think it is easier to be maintained there.

On another note, if this information is useful (yes since we needed at runtime) then may be we should consider put this information in the datatype structure and compute it during Type_commit time. Then since it is out-of time critical places, we can afford for more accurate stat if necessary.

raffenet replied:
I think `max_contig_blocks` only pertains to a single instance of the datatype. @pavanbalaji is that correct? I have no problem moving to `mpir_datatype.h`.

pavanbalaji replied:
Yes.  For example, if multiple datatypes are all contiguous, the total contig count should still be 1.  Similarly, if the end of one noncontig datatype and the start of the next are contiguous, those can be combined too.  Dataloop does those optimizations, though yaksa doesn’t.

FYI, your check for INT_MAX should be based on is_contig, not on BUILTIN datatypes.  Otherwise you’ll miss other contiguous types, such as dup of INT.

hzhou replied:
@pavanbalaji Does `max_contig_blocks` return the same as `MPIR_Typerep_get_iov_len`?

pavanbalaji replied:
Typerep takes a count argument, and max_contig_block is per datatype.  So they are not equivalent.  But max_contig_block * count >= what Typerep returns.

hzhou replied:
Thanks, it means I shouldn't use `MPIR_Typerep_get_iov_len` to calculate density.

pavanbalaji replied:
Actually, the opposite.  Typerep is the only accurate way to calculate the density, but it is more expensive to use than max_contig_block.  In this case, because an approximation is sufficient, max_contig_block is probably fine.

pavanbalaji replied:
To be clear: it’s more expensive for dataloop, but not for yaksa.

raffenet replied:
@pavanbalaji what is your suggestion for packing threshold? This PR uses 128 bytes per segment. Shawn mentioned using something like 16K(?) in the GPU work, but that probably assumes higher data movement cost.

pavanbalaji replied:
@raffenet `ch3` uses 16K, which is why @shawnccx picked that value.  Of course, it needs to be tuned, but 16K might be good starting point for now.

hzhou replied:
I don't think the default threshold matters for now. Either 128 or 16K is away from the extreme cases that we are trying to fix, so both or any should work as default. If we really care about the default performance, we need run measurements/study and then judge from the data.

pavanbalaji replied:
This PR should either run tests and find a good threshold or pick one that we have been using for `ch3`.  A different arbitrary number doesn't make sense.

raffenet replied:
Got it. I will use `MPIR_CVAR_CH4_IOV_DENSITY_MIN`, which is based on `ch3` from what I can tell.

pavanbalaji replied:
Sounds good for now.  In the future, we might have different thresholds, at least for CPUs vs. GPUs, but this is fine for now.

raffenet commented:
test:mpich/ch4/ofi

raffenet commented:
test:mpich/ch4/ofi

raffenet commented:
I believe PUT and GET are done now (minus tweaks to the packing threshold). I'm going to propose reviewing/merging this PR now, and doing a separate one for ACCUMULATE and GET_ACCUMULATE.

pavanbalaji commented:
@raffenet in `ch4/ofi: Add datatype density macro`, please see my comment above about not using `IS_BUILTIN`.

pavanbalaji commented on src/mpid/ch4/netmod/ofi/ofi_rma.h:
```diff
@@ -384,14 +384,114 @@ static inline int MPIDI_OFI_do_put(const void *origin_addr,
+    MPI_Aint num_iov, actual_iov_bytes;
+    int target_iov_len;
+    MPIR_Typerep_iov_len(target_count, target_datatype, target_bytes, &num_iov);
+    target_iov = MPL_malloc(sizeof(struct iovec) * num_iov, MPL_MEM_RMA);
```
Do we need to limit the size of the IOV buffer here?  In the worst case, we'd be allocating `target_bytes / MPIR_CVAR_CH4_OFI_IOVEC_MIN_SEG_SIZE` number of IOV elements.  Is that too large?

raffenet replied:
It could be large, yes. We could roughly allocate the number of IOVs that describe `max_msg_size` worth of data and then reuse them. That will require invoking Typerep multiple times with different offsets. Is that OK? I was originally trying to limit to a single Typerep call in each path.

pavanbalaji replied:
Here's the issue with IOVs: the right way to do this would be to have a limit on the memory you'd allocate.  So, for example, you can say you'll allocate a max of 16K IOV elements.  You'd get the first 16K IOV elements, issue them, get the next 16K IOV elements and so on.  This is not expensive to create.  This model is supported by yaksa, but not by Typerep.

Typerep only allows you to give offsets based on bytes, and not the number of IOV elements.  So, we'd need to calculate the number of bytes the first 16K IOV elements correspond to and then give that as an offset to Typerep.  Typerep would then convert those bytes back to IOV element count and return the next 16K elements.  This conversion is the expensive part.

One solution is to add an IOV-element-offset version to the Typerep API.  This would make yaksa take the fast path and dataloop take the slow path, which is the opposite of what we have today.  I believe such an API is more natural for what MPICH would need to use it as anyway.

pavanbalaji replied:
FYI, with the new API, pipelining it will not add much overhead.

pavanbalaji commented on src/mpid/ch4/netmod/ofi/ofi_rma.h:
```diff
@@ -384,14 +384,114 @@ static inline int MPIDI_OFI_do_put(const void *origin_addr,
+        MPI_Aint actual_pack_bytes;
+        req->noncontig.pack_buffer = MPL_malloc(origin_bytes, MPL_MEM_BUFFER);
+        MPIR_Typerep_pack(origin_addr, origin_count, origin_datatype, 0, req->noncontig.pack_buffer,
+                          origin_bytes, &actual_pack_bytes);
```
Do we need to limit the temporary buffer size here?  This can be as large as `origin_bytes`.

raffenet replied:
Same issue here. If multiple Typerep calls at OK, should be straightforward change to make.

pavanbalaji replied:
Using multiple Typerep calls here is not a problem.  It's been designed to use it in this manner.

raffenet replied:
So if we want to save memory, we would have to wait for each operation to complete before reloading the pack buffer. Perhaps we just fallback to AM if origin density is low and `origin_bytes > max_msg_size`?

pavanbalaji replied:
Would the AM code need an extra copy?  If yes, that'd be expensive for large messages.

raffenet replied:
Technically no, since ch4/ofi AM implementation uses FI_READ for payload transfer. But I'm going to need to abstract partial data operations anyway, so I might as well build it out for this, too.

raffenet commented:
> @raffenet in `ch4/ofi: Add datatype density macro`, please see my comment above about not using `IS_BUILTIN`.

Fixed in this version. Also moved the macro to `mpir_datatype.h`.

raffenet commented:
test:mpich/ch4/ofi

raffenet commented:
Converted this PR back to draft status. No-pack MPI_PUT implementation is added for testing purposes. Packing version still under development.
test:mpich/ch4/ofi

pavanbalaji commented:
@raffenet Can you punt the pack/unpack path to the AM layer, so the remaining patches can go in sooner?

pavanbalaji commented:
> @raffenet Can you punt the pack/unpack path to the AM layer, so the remaining patches can go in sooner?

I agree that the pack/unpack path is still needed, but that can perhaps be a separate PR.

raffenet commented:
> > @raffenet Can you punt the pack/unpack path to the AM layer, so the remaining patches can go in sooner?
> 
> I agree that the pack/unpack path is still needed, but that can perhaps be a separate PR.

OK, will do. Let me update for MPI_GET nopack path. Should be ready soon.

raffenet commented:
test:mpich/ch4/ofi

raffenet commented:
test:mpich/ch4/ofi

raffenet commented:
I think the only remaining discussion item is the IOV allocation limit. I set this at an arbitrary 1024 for testing. What is a good value for this? Is there some pre-existing limit elsewhere in MPICH?

pavanbalaji commented:
> I think the only remaining discussion item is the IOV allocation limit. I set this at an arbitrary 1024 for testing. What is a good value for this? Is there some pre-existing limit elsewhere in MPICH?

I think there are two potential IOVs: the origin IOV and the target IOV.  The origin IOV is simpler, because whatever is returned from yaksa/typerep can directly be fed to OFI.  The target IOV is more complicated.  We get a `struct iovec` from typerep, but it needs to be converted to `fi_rma_iov` for OFI.  The former is 16 bytes per element, the latter is 24 bytes per element.  So around 40 bytes.  I'd say, it's OK to have a limit of 16K IOV elements of each type.  That'd come to a maximum of around 1MB.

We should make these CVARs though.

raffenet commented:
test:mpich/ch4/ofi

raffenet commented:
https://github.com/pmodels/mpich/pull/4475/commits/0051a9f8ab6f4ed93cca21c39042a0be41932075 adds CVAR for iovec allocation limit. Now defaults to 16K.

raffenet commented:
Oh, let me update that comment...

raffenet commented:
test:mpich/ch4/ofi

pavanbalaji commented on src/include/mpir_datatype.h:
```diff
@@ -267,6 +267,21 @@ void MPIR_Datatype_get_flattened(MPI_Datatype type, void **flattened, int *flatt
+            (density_) = size_ / dt_ptr_->max_contig_blocks;	\
+        }                                                       \
+    } while (0)
+
```
You this be emacs indented?

raffenet replied:
Fixed.

pavanbalaji commented on src/mpid/ch4/netmod/ofi/ofi_rma.h:
```diff
@@ -176,6 +176,136 @@ MPL_STATIC_INLINE_PREFIX void MPIDI_OFI_sigreq_complete(MPIR_Request ** sigreq)
+        (void *) (MPIDI_OFI_winfo_base(win, target_rank) +
+                  (target_disp * MPIDI_OFI_winfo_disp_unit(win, target_rank)));
+    MPIR_Typerep_to_iov_offset(target_base, target_count, target_datatype, target_iov_offset,
+                               target_iov, max_iov_len, &target_iov_len);
```
I'm confused here.  Doesn't this need an extra loop to refill the IOV vector and issue the operations for cases where `total_target_iov_len` > `MPIR_CVAR_CH4_OFI_RMA_IOVEC_MAX`.

pavanbalaji replied:
For testing purposes, it'll be good to add a hack patch that temporarily sets the default IOV length to a smaller number (e.g., 1024), so that path can be tested too.  We'd obviously change it back to 16384 for the actual merge.

raffenet replied:
Yes, that happens in the while loop below. These could probably be unified to avoid the duplication.

pavanbalaji replied:
Oh, I see.  Yeah, it's a bit confusing to have one IOV load here and the rest down in the loop.  It might be better to have it in one place.

pavanbalaji commented:
@raffenet   It'll be good to rebase this to master when you push new commits, so as to keep the history as linear as possible.

raffenet commented:
> @raffenet It'll be good to rebase this to master when you push new commits, so as to keep the history as linear as possible.

I've rebased every time I've pushed today. I think Github notes if you are even a single commit out of date? It's not very easy to keep up these days.

raffenet commented:
test:mpich/ch4/ofi

raffenet commented on src/mpid/ch4/netmod/ofi/ofi_rma.h:
```diff
@@ -176,60 +176,134 @@ MPL_STATIC_INLINE_PREFIX void MPIDI_OFI_sigreq_complete(MPIR_Request ** sigreq)
+    int i = 0, j = 0;
+    size_t msg_len;
+    while (i < origin_iov_len && j < target_iov_len) {
+        if (i == 0) {
```
iovec population now happens in a single place. This loop is kind of complex, but at least it condenses the typerep calls now.

raffenet commented:
test:mpich/ch4/ofi

raffenet commented:
test:mpich/ch4/ofi

raffenet commented:
Indexing logic was made a little simpler using % operator. I also added a iov load helper in this version.

pavanbalaji commented on src/mpid/ch4/netmod/ofi/ofi_rma.h:
```diff
@@ -176,6 +176,127 @@ MPL_STATIC_INLINE_PREFIX void MPIDI_OFI_sigreq_complete(MPIR_Request ** sigreq)
+        MPI_Aint outlen;
+        MPIR_Typerep_to_iov_offset(buffer, count, datatype, *offset, iov, max_len, &outlen);
+        *offset += outlen;
+    }
```
This logic is a lot more cleaner and easier to follow that the previous draft.  Perhaps the variables can be a bit more descriptive?  How about:

`done` -> `current_iov_offset`
`offset` -> `loaded_iov_offset`

raffenet replied:
Updated in latest version.

raffenet commented:
test:mpich/ch4/ofi

pavanbalaji commented:
Would you mind also testing it with a hack patch that changes `--with-datatype-engine` to `auto`, so the `yaksa` code gets tested too?

pavanbalaji commented on src/mpid/ch4/netmod/ofi/ofi_rma.h:
```diff
@@ -176,6 +176,128 @@ MPL_STATIC_INLINE_PREFIX void MPIDI_OFI_sigreq_complete(MPIR_Request ** sigreq)
+        } else {
+            j++;
+        }
+    }
```
```suggestion
    }
    MPIR_Assert(i == total_origin_iov_len);
    MPIR_Assert(j == total_target_iov_len);
```

raffenet replied:
Added in newest version.

raffenet replied:
Added in latest version.

raffenet commented:
test:mpich/ch4/most

raffenet commented:
test:mpich/ch4/ofi

raffenet commented:
Assuming tests are successful, will pop off the HACK patches and merge.

raffenet commented:
There's a bunch of failures in the latest test run that I believe are related to yaksa. https://jenkins-pmrs.cels.anl.gov/job/mpich-review-ch4-ofi/1230/

If I revert the yaksa patch and run some tests by hand, they pass.

raffenet commented:
@pavanbalaji are the yaksa failures as expected? Just double checking before merge.

pavanbalaji commented:
@raffenet Not expected, but I'll work on them separately from this PR.  No need to hold up this PR for them.
